#!/bin/sh
#PBS -l mem=16gb,nodes=1,ppn=1,walltime=72:00:00
#PBS -m abe
#PBS -M konox006@umn.edu
#PBS -q lab

#   This is the ANGSD script from RILab's BigD study,
#   modified for running on MSI by TomJKono.

#   Last Modified: 2014-06-11
#   CHANGES
#       2014-06-11
#           - Cleaned uop the script, fixed some directory names for pointing
#             to directories in our MSI shared space.

#   First we specify values for all of our ANGSD analyses
#   The directory of our reference sequence
REF_DIR=/home/morrellp/shared/References/Reference_Sequences/Barley/Morex/
#   This sequence is the pseudo-scaffolds from Martin
REF_SEQ=131012_morex_pseudoscaffolds.fasta
#   Since ANGSD's version is in the directory name, we reference them both
#   in this way
ANGSD_VERSION=0.602
ANGSD_DIR=/home/morrellp/shared/Software/angsd${ANGSD_VERSION}
#   Variable stores the division of the data we are looking at
#   e.g., wild, landrace or cultivar
#   We have a separate directory for associated data files
#   In this case, the TAXON_LIST variable contains the bam filelist file for
#   all the samples we are analyzing
TAXON=cultivar
TAXON_LIST=data/${TAXON}_samples.txt
TAXON_INBREEDING=data/${TAXON}_F.txt
#   The number of individuals in the taxon we are analyzing
#   We use an embedded command to do this
#   ( wc -l < FILE will return just the line count of FILE,
#   rather than the line count and the filename. More efficient than piping
#   to a separate 'cut' process!)
N_IND=`wc -l < ${TAXON_LIST}`
#   For ANGSD, the actual sample size is twice the number of individuals, since
#   each individual has two chromosomes. The individual inbreeding coefficents
#   take care of the mismatch between these two numbers
N_CHROM=`expr 2 \* ${N_IND}`

#   This first command estimates the site frequency spectrum
#   Options:
#       -bam [FILE]
#           List of bam file paths in [FILE]
#       -out [FILE]
#           Output SFS to [FILE]
#       -doSaf [1|2|3|4]
#           Model used to determine genotype likelihoods
#               1: Assume HWE
#               2: Take into account individual inbreeding coefficients.
#                  Requires the -indF argument (We will use this one)
#               3: Calculate posterior allele frequency dist. from the file
#                  generated by -doSaf 1. (Seems circular?)
#               4: Calculate posterior probabilities given genotype probs
#                  from another source. Requires -beagle.
#       -uniqueOnly [0/1]
#           0: Do not use only unique reads
#           1: Use only uniquely mapped reads (we use this one)
#       -anc [FILE] 
#           ancestral sequence (see trip.sh script for how this is generated)
#           This is a FASTA file
#       -minMapQ [INT]
#           Minimum mapping quality of reads to accept. MapQ is a Phred-scaled
#           probability of a read having an incorrect mapping position
#       -minQ [INT]
#           Minimum base quality to be used in calculations. This is the Phred
#           scaled quality score that comes with the FASTQ reads
#       -minInd [INT]
#           Important parameter; specifies how many individuals are allowed to
#           have missing data at a site. Sites above this threshhold will be
#           filtered from the analysis
#       -setMaxDepth [INT] 
#           Maximum depth to consider, good for filtering out highly repetitive
#           regions from analysis
#       -baq [0/1]
#           Use the Base Alignment Quality score feature
#           See http://samtools.sourceforge.net/mpileup.shtml and Li 2011
#           http://bioinformatics.oxfordjournals.org/content/27/8/1157.long
#           for more information. 
#               1: Use BAQ, which tries to account for false positive variation
#                  due to misalignment across indels. It assigns a Phred-scaled
#                  probability that the base is misaligned
#               0: Do not use BAQ
#       -GL [1|2|3|4]
#           Model used to calculate genotype likelihoods
#           Explained: http://popgen.dk/angsd/index.php/Genotype_Likelihoods
#               1: SAMTools
#               2: GATK (We will use this one)
#               3: SOAPsnp
#               4: SYK
#       -r [RANGE]
#           Operate only on [RANGE], given in standard genomic coordinate form
#           i.e., chr:start-stop (morex_contig_1:1-100)
#       -P [INT]
#           Use [INT] threads
#       -indF
#           individiual inbreeding coefficient. for inbred lines just make a
#           files of "1" on each line for each bamfile. otherwise use ngsF
#           to estimate (see inbreeding.sh script)
#       -r [REGION] / [FILE]
#           operate over [REGION] or the regions denoted in [FILE]
#           One region per line, specified in standard genomic coordinate
#           format, chr:start-stop
#           e.g., morex_contig_1:1-100

#   We set operation-specific parameters here.
#   Hopefully these variable names make sense . . .
DO_SAF=2
UNIQUE_ONLY=1
#   Bulbosum sequence for ancestral state
ANCESTRAL=data/Bulbosum.fa.gz
MIN_BASEQUAL=20
BAQ=1
MIN_IND=1
GT_LIKELIHOOD=2
MIN_MAPQ=30
N_CORES=32
DO_MAJORMINOR=1
DO_MAF=1
#   REGIONS value yanked from JRI script
#   will update when we get our regions of interest
REGIONS="morex_contig_43476:1310-3262"

#   Now we actually run the command, this creates a binary file that contains the prior SFS
${ANGSD_DIR}/angsd \
    -bam ${TAXON_LIST}\
    -out ${TAXON}_SFSOut\
    -indF ${TAXON_INBREEDING}\
    -doSaf ${DO_SAF}\
    -uniqueOnly ${UNIQUE_ONLY}\
    -anc ${ANCESTRAL}\
    -minMapQ ${MIN_MAPQ}\
    -minQ ${MIN_BASEQUAL}\
    -nInd ${N_IND}\
    -minInd ${MIN_IND}\
    -baq ${BAQ}\
    -ref ${REF_DIR}/${REF_SEQ}\
    -GL ${GT_LIKELIHOOD}\
    -P ${N_CORES}\
    -doMajorMinor $DO_MAJORMINOR\
    -doMaf $DO_MAF\
    -r ${REGIONS}

#   This next command estimates the SFS
#   Notes from JRI:
#       -fold option seems deprecated
#       the .saf output file is prior on SFS, probably
#       number of chromosomes is 2x number of samples
#       .em.ml output file is the final estimate of SFS
#       final SFS estimate will be ln(P) where P is the probability of the
#       values of SFS from 0 to the sample size. (will give n+1 values total)
#       for polymorphic sites, ignore the first and last numbers
#       first number is ln(proportion fixed for ancestral allele)
#       last number is ln(proportion fixed for derived allele)
${ANGSD_DIR}/misc/emOptim2\
    ${TAXON}_SFSOut.saf\
    ${N_CHROM}\
    -P ${N_CORES}\
    > ${TAXON}_DerivedSFS

#   Next, use the SFS to calculate various diversity stats
#   Options:
#       -doThetas [0|1]
#           0: Do not estimate diversity
#           1: Calculate nucleotide diversity, thetaH, thetaL, wattersons theta
#       -pest [FILE]
#           Use the esitmated SFS in [FILE]
#           This is the output file from the above command
#
#   The .thetas output file will have seven columns, and contain data for every
#   site in the input, including monomorphic sites. The values in the columns
#   are powers of 10 for the estimate of diverstiy at that site
#   Example output from JRI:
#
#Chromo Pos    Watterson  Pairwise    thetaSingleton  thetaH      thetaL
#10     3370   -8.664392  -10.223986  -7.289949       -13.844801  -10.890724
#10     3371   -8.822116  -10.395367  -7.431857       -14.041094  -11.062746
#10     3372   -8.840759  -10.415518  -7.448764       -14.064022  -11.082968
#10     26926  -1.480456  -0.671793   -211.328599     -0.694813   -0.683237
#
#   In this case, the first site is probably not polymorphic; the estimate of
#   theta is ~10^(-10). The last site is probably polymorphic; the estimate of
#   theta is about 10^(-0.67) ~ 0.22. This should be similar to the MAF at that
#   position, too.
#   I think for getting a locus average, we can just sum the values and divide
#   by the locus length?
#   It's also odd that there are no variances for these estimates - again
#   perhaps we can do that on a per-locus basis?

#   Set some operation-specific parameters here
DO_THETAS=1
PEST=${TAXON}_DerivedSFS

#   And actually run the command
${ANGSD_DIR}/angsd\
    -bam ${TAXON_LIST}\
    -out ${TAXON}_Divsersity\
    -indF ${TAXON_INBREEDING}\
    -doSaf ${DO_SAF}\
    -doThetas ${DO_THETAS}\
    -uniqueOnly ${UNIQUE_ONLY}\
    -anc ${ANCESTRAL}\
    -minMapQ ${MIN_MAPQ}\
    -minQ ${MIN_BASEQUAL}\
    -nInd ${N_IND}\
    -minInd ${MIN_IND}\
    -baq ${BAQ}\
    -ref ${REF_DIR}/${REF_SEQ}\
    -GL ${GT)LIKELIHOOD}\
    -P ${N_CORES}\
    -pest ${PEST}\
    -doCounts\
    -doDepth\
    -dumpCounts\
    ${REGIONS}

#   Next is to calculate Tajima's estimator
#   This makes a somewhat bedfile-like output including Tajima's D and others
${ANGSD_DIR}/misc/thetaStat make_bed\
    ${TAXON}_Diversity.thetas.gz\
    ${TAXON}_Tajimas

#   This calculates Tajima's D (And other statistics) in sliding windows
#   Options:
#       -nChr [INT]
#           [INT] number of chromosomes
#       -step [INT]
#           Step [INT] basepairs between windows
#       -win [INT]
#           Windows are [INT] basepairs wide
#   This will output a .pestPG file which will contain 14 column file
#   The columns are
#       1: (indexStart,indexStop)(posStart,posStop)(regStat,regStop)
#       2: chrname - chromosome name
#       3: wincenter - center of window
#       4: tW - Watterson's Theta
#       5: tP - Pairwise Theta (pi?)
#       6: tF - Fu and Li?
#       7: tH - Fay?
#       8: tL - Fay?
#       9: tajD - Tajima's D
#       10: fulif - Fu and Li's F
#       11: fuliD - Fu and Li's D
#       12: fayH - Fay and Wu?
#       13: zengsE - Zheng's E
#       14: numSites - Number of sites in window
#   There will be one row per window
#   For more information see:
#       http://popgen.dk/angsd/index.php/Tajima#.thetas.gz.pestPG

#   Set some operation-specific parameters
WINDOW_SIZE=1000
WINDOW_STEP=500

#   And run the analysis!
${ANGSD_DIR}/misc/thetaStat do_stat\
    ${TAXON}_Tajimas\
    -nChr ${N_CHROM}\
    -win ${WINDOW_SIZE}\
    -step ${WINDOW_STEP}
